This PR fixes two bugs in the protected _set_joint_torque and _get_joint_torques of the DQ_CoppeliaSimInterfaceZMQ class

juanjqo opened the pull request:
![](https://img.shields.io/badge/Tests-developer%20workflow-orange)![](https://img.shields.io/badge/Ubuntu%2024.04.5%20LTS%20(x64)-tested-passing)

@dqrobotics/developers 


Hi @mmmarinho,


This PR fixes two bugs in the protected _set_joint_torque and _get_joint_torques methods of the `DQ_CoppeliaSimInterfaceZMQ` class. Specifically, this PR fixes the bug #9. Furthermore,  I fixed a bug in the method _set_joint_torque. This bug consists in that only positive torques are set in the joint. In the legacy API, we used to set always positive torques, in which the desired direction of motion were set by a high signed velocity. This is not working anymore in the ZMQ remote API. Instead, we can now define the desired direction of motion using the sign of the torque using just [simSetJointTargetForce()](https://manual.coppeliarobotics.com/en/regularApi/simSetJointTargetForce.htm)

I tested the modifications using the same example I used in my last Python PR. I opened a [PR](https://github.com/dqrobotics/coppeliasim-scenes) to propose the new scene.

Kind regards, 

Juancho

### Tests

```cpp
#include <iostream>
#include <dqrobotics/DQ.h>
#include <dqrobotics/utils/DQ_Constants.h>
#include <dqrobotics/interfaces/coppeliasim/DQ_CoppeliaSimInterfaceZMQ.h>


using namespace DQ_robotics;
using namespace Eigen;

int main()
{
    auto cs = DQ_CoppeliaSimInterfaceZMQ();

    try{
        // Establishes the connection with CoppeliaSim.
        cs.connect("localhost", 23000, 800);

        // Set the stepping mode.
        // Check more in https://manual.coppeliarobotics.com/en/simulation.htm
        cs.set_stepping_mode(true);

        cs.start_simulation();

        // Define some parameters to compute varying-time trajectories
        double a{1};
        double freq{0.1};
        double time_simulation_step{0.05};

        for (size_t i=0;i<300;i++)
        {
            double t = i*time_simulation_step;

            // Define a varying-time position based on the Lemniscate of Bernoulli
            DQ p = (a*cos(t)/(1 + pow(sin(t),2)))*i_ + (a*sin(t)*cos(t)/(1 + pow(sin(t),2)))*j_;

            // Define a varying-time orientation
            DQ r = cos(2*pi*freq*t) + k_*sin(2*pi*freq*t);

            // Built a varying-time unit dual quaternion
            DQ x = r + 0.5*E_*p*r;

            // Set the object pose in CoppeliaSim
            cs.set_object_pose("/coffee_drone", x);

            // Read the pose of an object in CoppeliaSim to set the pose of
            // another object with a constant offset.
            DQ xread = cs.get_object_pose("/coffee_drone");
            DQ xoffset = 1 + 0.5*E_*(0.5*i_);
            DQ xnew = xread*xoffset;
            cs.set_object_pose("/Frame_x", xnew);


            // Set the target position of the first joint of the UMIRobot arm
            VectorXd target_position = (VectorXd(1) <<sin(2*pi*freq*t)).finished();
            std::vector<std::string> umi_joint = {"UMIRobot/UMIRobot_joint_1"};
            cs.set_joint_target_positions(umi_joint, target_position);

            // Set the target position of the third joint of the UR5 robot
            std::vector<std::string> ur5_joint = {"UR5/link/joint/link/joint"};
            cs.set_joint_target_positions(ur5_joint, target_position);

            // Set the target velocity of the first joint of the Franka Emika Panda
            VectorXd target_velocity = (VectorXd(1) <<2*pi*freq*cos(2*pi*freq*t)).finished();
            std::vector<std::string> franka_joint = {"Franka/joint"};
            cs.set_joint_target_velocities(franka_joint, target_velocity);

            // Set the torque of the Dummy Robot
            VectorXd torque = (VectorXd(1) <<sin(2*pi*freq*t)).finished();
            std::vector<std::string> dummy_joint = {"Revolute_joint"};
            cs.set_joint_torques(dummy_joint, torque);



            // Set the target velocities of the Pioneer wheels
            std::vector<std::string> pioneer_rjoint = {"PioneerP3DX/rightMotor"};
            std::vector<std::string> pioneer_ljoint = {"PioneerP3DX/leftMotor"};
            cs.set_joint_target_velocities(pioneer_rjoint, (VectorXd(1) << 0.1).finished());
            cs.set_joint_target_velocities(pioneer_ljoint, (VectorXd(1) << 0.2).finished());


            // Trigger a simulation step in CoppeliaSim
            cs.trigger_next_simulation_step();

            auto torque_read = cs.get_joint_torques(dummy_joint );
            std::cout<<torque<<"  , "<<torque_read<<std::endl;
        }
        cs.stop_simulation();

    } catch (std::exception& e) {

        std::cout<<e.what()<<std::endl;
        cs.stop_simulation();
    }
}

```

Output:
```shell
[...]
-0.125333  , -0.125333
-0.0941083  , -0.0941083
-0.0627905  , -0.0627905
-0.0314108  , -0.0314108
-2.44929e-16  , -2.44929e-16
0.0314108  , 0.0314108
0.0627905  , 0.0627905
0.0941083  , 0.0941083
0.125333  , 0.125333
0.156434  , 0.156434
[...]
```

juanjqo committed:
[DQ_CoppeliaSimInterfaceZMQ] Fixed bugs in _set_joint_torque and fixed the bug reported by Frederico concerning the _get_joint_torque method.

mmmarinho commented on src/dqrobotics/interfaces/coppeliasim/DQ_CoppeliaSimInterfaceZMQ.cpp:
```diff
@@ -765,7 +756,7 @@ void DQ_CoppeliaSimInterfaceZMQ::set_joint_torques(const std::vector<std::string
 {
     _check_client();
-    return _ZMQWrapper::get_sim()->getJointForce(handle);
+    return -_ZMQWrapper::get_sim()->getJointForce(handle);
```
@juanjqo I think GitHub interface isn't showing everything, but if there's a `-` sign here it's probably best to have some additional explanation.

juanjqo replied:
Hi @mmmarinho.  I agree. I'll add the comments. I think there is a bug in the `getJointForce` ZMQ remote API method. I open an [issue](https://forum.coppeliarobotics.com/viewtopic.php?p=41690#p41690) in the CoppeliaSim Forum.

mmmarinho reviewed (changes requested):
@juanjqo  Could you please let me know if this is ok? If so I can merge.

mmmarinho commented:
@juanjqo 

To better explain my question. It's better to understand what's the reason for us needing this negative signal. A little reading makes me think this is related to the physics engine: does `get_force` mean applied to the joint or acted by the joint? As long as we understand the reasons and this doesn't vary among engines I think we're good to go.

https://manual.coppeliarobotics.com/en/regularApi/simGetJointForce.htm

> Retrieves the force or torque applied to a joint along/about its active axis. This function retrieves meaningful information only if the joint is prismatic or revolute, and is dynamically enabled. With the Bullet, MuJoCo and Newton engine, this function returns the force or torque applied to the joint motor (torques from joint limits are not taken into account). With the ODE and Vortex engine, this function returns the total force or torque applied to a joint along/about its z-axis

juanjqo commented:
@mmmarinho I'm going to run some tests and wait for Coppelia's support to understand better what is happening. In the meantime, I'm going to convert this PR into a draft.

mmmarinho commented:
Thanks, @juanjqo!

> Hi @mmmarinho. I agree. I'll add the comments. I think there is a bug in the `getJointForce` ZMQ remote API method. I open an [issue](https://forum.coppeliarobotics.com/viewtopic.php?p=41690#p41690) in the CoppeliaSim Forum.

The response to the issue you opened in the forum was (copying here so that we keep track)

> [sim.getJointForce](https://www.coppeliarobotics.com/helpFiles/en/regularApi/simGetJointForce.htm) is not the same as [sim.getJointTargetForce](https://www.coppeliarobotics.com/helpFiles/en/regularApi/simGetJointTargetForce.htm), even if related.
The sign depends on what the reference is. If the joint has its Z axis up and is holding a mass above the ground, then the mass exerts a negative force onto the joint (into the joint's negative Z vector direction)

So, this matches my suspicion that it was the force enacted on the joint, not the force enacted by the joint. Hence, the sign being negative makes sense. No doubt you're aware of this, but nonetheless the method itself needs a review so that you're getting what you believe you are, and this doesn't cause issues anywhere else.

juanjqo committed:
[DQ_CoppeliaSimInterfaceZMQ.cpp] Improved the documentation of the set/get joint torque.

juanjqo commented:
> Thanks, @juanjqo!
> 
> > Hi @mmmarinho. I agree. I'll add the comments. I think there is a bug in the `getJointForce` ZMQ remote API method. I open an [issue](https://forum.coppeliarobotics.com/viewtopic.php?p=41690#p41690) in the CoppeliaSim Forum.
> 
> The response to the issue you opened in the forum was (copying here so that we keep track)
> 
> > [sim.getJointForce](https://www.coppeliarobotics.com/helpFiles/en/regularApi/simGetJointForce.htm) is not the same as [sim.getJointTargetForce](https://www.coppeliarobotics.com/helpFiles/en/regularApi/simGetJointTargetForce.htm), even if related.
> > The sign depends on what the reference is. If the joint has its Z axis up and is holding a mass above the ground, then the mass exerts a negative force onto the joint (into the joint's negative Z vector direction)
> 
> So, this matches my suspicion that it was the force enacted on the joint, not the force enacted by the joint. Hence, the sign being negative makes sense. No doubt you're aware of this, but nonetheless the method itself needs a review so that you're getting what you believe you are, and this doesn't cause issues anywhere else.

@mmmarinho thanks. The behaviour of the method matches your description and Coppelia's reply. I updated the documentation and added comments explaining why we need to multiply by -1. This description is already on the [Matlab version](https://github.com/dqrobotics/matlab-interface-coppeliasim-zmq/blob/cd5bc400040005e1cf0ce528265891d206099d44/interfaces/coppeliasim/DQ_CoppeliaSimInterfaceZMQ.m#L419).
